### PR TITLE
Rename `Reference` to `RefCounted`

### PR DESCRIPTION
**Pull request by RandomShaper**
This implements https://github.com/godotengine/godot-proposals/issues/652.

After the big find'n'replace, I've carefully checked every change, adapting some variable names to what made sense and reverting those replacements that were pointless.

I'm only a bit unsure about some changes in the `.cs` files (CC @neikeq).

### CODE REVIEW

**Review by fire — APPROVED** [review]
Not a code review.

Design review!

As a member of the Import and Usability teams, I think this makes things more clear and causes less confusion for the concepts of Reference. However, I would wait for the Engine team to review.

**Review by neikeq — APPROVED** [review]
LGTM

I don't like how in some variable names `ref` was changed to `rc` because I have seen RC used in other places for "resource".

**Comment by DarkMessiah** [comment]
And in GDScript the class will also be called RefCounted instead of Reference?

**Comment by RandomShaper** [comment]
> And in GDScript the class will also be called RefCounted instead of Reference?

Yes.

**Comment by akien-mga** [comment]
Approved in PR review meeting, just needs a rebase.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: modules/mono/glue/base_object_glue.cpp [hunk 1]
```diff
@@ -81,13 +81,13 @@ void godot_icall_Object_Disposed(MonoObject *p_obj, Object *p_ptr) {
...
+	// This is only called with RefCounted derived classes
+	CRASH_COND(!Object::cast_to<RefCounted>(p_ptr));
 #endif
 
-	Reference *ref = static_cast<Reference *>(p_ptr);
+	RefCounted *rc = static_cast<RefCounted *>(p_ptr);
 
-	if (ref->get_script_instance()) {
+	if (rc->get_script_instance()) {
 		CSharpInstance *cs_instance = CAST_CSHARP_INSTANCE(ref->get_script_instance());
```
Comment (rasmuskr) [inline modules/mono/glue/base_object_glue.cpp:91]: These changes does not seem to be able to compile when enabling mono as ref looks have been renamed to rc

**Review by rasmuskr — COMMENTED** [review]
This change seems to have compile errors when compiling with mono. (only tested on windows though)

File: modules/mono/glue/base_object_glue.cpp (hunk 1 above)
Comment (RandomShaper) [reply modules/mono/glue/base_object_glue.cpp:91]: #49562 :)

**Pull request merged**